From 815435cfcc5993db813df09ffcb22f9b0e3a5357 Mon Sep 17 00:00:00 2001
From: Marco Hunsicker <marco@…>
Date: Fri, 14 Feb 2014 19:19:48 +0100
Subject: [PATCH] Bugfix: Don't set input focus to window that has been closed

This patch adjusts the point in time the focus is transferred when
a window is closed to avoid having the input focus set (again) to
the already detached window.

Fixes #1185
---
 src/tree.c |   40 +++++++++++++++++++++-------------------
 1 file changed, 21 insertions(+), 19 deletions(-)

diff --git a/src/tree.c b/src/tree.c
index 48ab163..7f28c3c 100644
--- a/src/tree.c
+++ b/src/tree.c
@@ -295,6 +295,27 @@ bool tree_close(Con *con, kill_window_t kill_window, bool dont_kill_parent, bool
         con_fix_percent(parent);
     }
 
+    if (next) {
+        if (was_mapped || con == focused) {
+            if ((kill_window != DONT_KILL_WINDOW) || !dont_kill_parent || con == focused) {
+                DLOG("focusing %p / %s\n", next, next->name);
+                if (next->type == CT_DOCKAREA) {
+                    /* Instead of focusing the dockarea, we need to restore focus to the workspace  */
+                    con_focus(con_descend_focused(output_get_content(next->parent)));
+                } else {
+                    if (!force_set_focus && con != focused)
+                        DLOG("not changing focus, the container was not focused before\n");
+                    else con_focus(next);
+                }
+            }
+            else {
+                DLOG("not focusing because we're not killing anybody\n");
+            }
+        } else {
+            DLOG("not focusing, was not mapped\n");
+        }
+    }
+
     /* Render the tree so that the surrounding containers take up the space
      * which 'con' does no longer occupy. If we don’t render here, there will
      * be a gap in our containers and that could trigger an EnterNotify for an
@@ -327,25 +348,6 @@ bool tree_close(Con *con, kill_window_t kill_window, bool dont_kill_parent, bool
         return true;
     }
 
-    if (was_mapped || con == focused) {
-        if ((kill_window != DONT_KILL_WINDOW) || !dont_kill_parent || con == focused) {
-            DLOG("focusing %p / %s\n", next, next->name);
-            if (next->type == CT_DOCKAREA) {
-                /* Instead of focusing the dockarea, we need to restore focus to the workspace */
-                con_focus(con_descend_focused(output_get_content(next->parent)));
-            } else {
-                if (!force_set_focus && con != focused)
-                    DLOG("not changing focus, the container was not focused before\n");
-                else con_focus(next);
-            }
-        }
-        else {
-            DLOG("not focusing because we're not killing anybody\n");
-        }
-    } else {
-        DLOG("not focusing, was not mapped\n");
-    }
-
     /* check if the parent container is empty now and close it */
     if (!dont_kill_parent)
         CALL(parent, on_remove_child);
-- 
1.7.10.4

